### PR TITLE
[Serialization] Implement serialization for @differentiable attribute.

### PR DESCRIPTION
**Pull request by dan-zheng**
Implement serialization/deserialization for all components of `@differentiable` attribute except the trailing where clause (which needs to be type-checked).

This is a necessary step for the `#adjoint` expression to look up `@differentiable` attributes declared on functions in other modules.

Addresses [SR-7977](https://bugs.swift.org/browse/SR-7977).

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: include/swift/Serialization/ModuleFormat.h [hunk 1]
```diff
@@ -1496,6 +1494,17 @@ namespace decls_block {
...
+  using DifferentiableDeclAttrLayout = BCRecordLayout<
+    Differentiable_DECL_ATTR,
+    BCFixed<1>, // Differentiation mode ('forward' or 'reverse').
+    IdentifierIDField, // Primal name.
+    DeclIDField, // Primal function declaration.
+    IdentifierIDField, // Adjoint name.
+    DeclIDField, // Adjoint function declaration.
+    BCArray<BCFixed<32>> // Differentiation parameters.
+  >;
+
```
Comment (rudkx) [inline include/swift/Serialization/ModuleFormat.h]: VERSION_MINOR in this file needs to be bumped.

Comment (dan-zheng) [reply include/swift/Serialization/ModuleFormat.h]: Thanks! Done in https://github.com/apple/swift/pull/17155/commits/d7ea8bc368c53b8e5fe9600306bff33123516eb1.

File: include/swift/Serialization/ModuleFormat.h [hunk 2]
```diff
@@ -55,7 +55,7 @@ const uint16_t VERSION_MAJOR = 0;
 /// describe what change you made. The content of this comment isn't important;
 /// it just ensures a conflict if two people change the module format.
 /// Don't worry about adhering to the 80-column limit for this line.
-const uint16_t VERSION_MINOR = 405; // SWIFT_ENABLE_TENSORFLOW: graph_op.
+const uint16_t VERSION_MINOR = 406 // SWIFT_ENABLE_TENSORFLOW: serialize @differentiable.
```
Comment (rxwei) [inline include/swift/Serialization/ModuleFormat.h]: Semicolon

File: lib/Serialization/Deserialization.cpp [hunk 3]
```diff
@@ -2528,6 +2528,51 @@ ModuleFile::getDeclCheckedImpl(DeclID DID, Optional<DeclContext *> ForcedContext
...
+          primalDecl = cast<FuncDecl>(getDecl(primalDeclId));
+        }
+        FuncSpecifier adjoint = { getIdentifier(adjointNameId), DeclNameLoc() };
+        FuncDecl *adjointDecl = cast<FuncDecl>(getDecl(adjointDeclId));
+
+        SmallVector<AutoDiffParameter, 4> parameters;
+        for (auto paramValue : paramValues) {
+          auto parameter = paramValue & 0x01
+            ? AutoDiffParameter::getSelfParameter(SourceLoc())
+            : AutoDiffParameter::getIndexParameter(SourceLoc(), paramValue>>1);
```
Comment (rxwei) [inline lib/Serialization/Deserialization.cpp]: `paramValue >> 1`

**Review by rxwei — APPROVED** [review]
Awesome work!

File: lib/Serialization/Deserialization.cpp (hunk 3 above)
Comment (dan-zheng) [reply lib/Serialization/Deserialization.cpp]: Done, minor refactoring to keep things within 80 cols.
I too love spaces around operators.

File: include/swift/Serialization/ModuleFormat.h (hunk 2 above)
Comment (dan-zheng) [reply include/swift/Serialization/ModuleFormat.h]: Done, thanks.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Pull request merged**